<a href="https://colab.research.google.com/github/thehapyone/Thesis_Project/blob/master/Entity_Embeddings_v6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Learned Embeddings
This program will attempt to create embeddings for the categorical features using the pipelines code developed.

In [95]:
from google.colab import drive
drive.mount('/content/drive', force_remount=False)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [0]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import copy as copy

In [0]:
import sys
sys.path.append('drive/My Drive/Thesis/Collab Notebooks/')

In [0]:
#%run drive/"My Drive"/Thesis/"Collab Notebooks"/category_pipeline.ipynb
#%load

In [0]:
# using the latest updated file
data = pd.read_csv("drive/My Drive/Thesis/data/Data_Engineering/Edited_Raw_Data/v3_Sorted_Raw_Db_with_RUL.csv", sep=',', low_memory=False)
#data = pd.read_csv("drive/My Drive/Thesis/data/Feature_Engineering/KNN_Imputed_Datasets/KNN_Imputed_k_40.csv", sep=',', low_memory=False)
#data = pd.read_csv("drive/My Drive/Thesis/data/Data_Engineering/Edited_Raw_Data/v2_Sorted_Database.csv", sep=',', low_memory=False)

In [100]:
data.shape

(19954, 449)

## Pipeline Extraction

In [0]:
# Applying the Extraction pipeline
# first stage of the pipeline
from extraction_pipeline import *
from category_pipeline import *

In [0]:
# create a backup
original_data = copy.copy(data)


In [103]:
'''
# remove duplicates
#data = removeDuplicates(data, debug=True)
data = data.drop(labels=['DELIVERY_DATE.1'], axis=1)
print ('After duplicates removal - ',data.shape)

# the only duplicate column discovered was - ['DELIVERY_DATE.1']
print ('After duplicates removal - ',data.shape)
# Remove zero variance
data = removeZeroVariance(data, debug=True)
print ('After Zero Variance removal - ',data.shape)

# drop redundant features
data = dropRedundantFeatures(data)
print ('After Redundant Drop removal - ',data.shape)

cols_left = data.columns
'''

"\n# remove duplicates\n#data = removeDuplicates(data, debug=True)\ndata = data.drop(labels=['DELIVERY_DATE.1'], axis=1)\nprint ('After duplicates removal - ',data.shape)\n\n# the only duplicate column discovered was - ['DELIVERY_DATE.1']\nprint ('After duplicates removal - ',data.shape)\n# Remove zero variance\ndata = removeZeroVariance(data, debug=True)\nprint ('After Zero Variance removal - ',data.shape)\n\n# drop redundant features\ndata = dropRedundantFeatures(data)\nprint ('After Redundant Drop removal - ',data.shape)\n\ncols_left = data.columns\n"

In [104]:
# date extractor
data = dateInfoExtractor(raw_data=data, debug=False)
print ('After date extractor - ',data.shape)

After date extractor -  (19954, 451)


In [105]:
# missing values analysis
#
print ("missing values rate before: ", 100 * (data.isnull().sum().sort_values(ascending=False) / data.shape[0]) )
data = fillMissingValues(data)
print ("missing values rate after: ", 100 * (data.isnull().sum().sort_values(ascending=False) / data.shape[0]) )


missing values rate before:  LX_PSC_P1FUO_NUMBER_OF_STARTER_ACTIVATION_D    15.590859
LX_PSC_P1FUN_NUMBER_OF_ENGINE_STARTS_DURING    14.628646
LX_PSC_P1MNF_COMPRESSOR_LOAD_PHASE_COUNTER      7.933246
LX_PSC_P1ASH_COOLANT_PUMP_NUMBER_OF_ACTIVAT     4.951388
LX_PSC_P1ASI_COOLANT_PUMP_ACTIVATION_TIME       4.951388
                                                 ...    
VFE_0006_VEHICLE_OPERATION_DIGIT2               0.000000
VFE_0005_VEHICLE_OPERATION_DIGIT1               0.000000
LX_PSC_P1BBY_TOTAL_FUEL_CONSUMPTION             0.000000
SLX_TOTAL_VEHICLE_DISTANCE_RULE_BASED           0.000000
DAY                                             0.000000
Length: 451, dtype: float64


/usr/local/lib/python3.6/dist-packages/sklearn/impute/_iterative.py:638: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.
  " reached.", ConvergenceWarning)


missing values rate after:  LX_PSC_P1MNF_COMPRESSOR_LOAD_PHASE_COUNTER     7.933246
LX_PSC_P1ASI_COOLANT_PUMP_ACTIVATION_TIME      4.951388
LX_PSC_P1ASH_COOLANT_PUMP_NUMBER_OF_ACTIVAT    4.951388
LX_PSC_P1I80_TOP_VALUE_OF_ENGINE_SPEED_IN_O    4.946377
X_PSC_P1ATS_OIL_DEGRADATION_LAST_OLC_HOURS     4.715846
                                                 ...   
SLX_TOTAL_VEHICLE_DISTANCE_RULE_BASED          0.000000
SLX_PSC_P1IPV_INJECTED_UREA_TOTAL_MASS         0.000000
SLX_PSC_P1BBY_TOTAL_FUEL_CONSUMPTION           0.000000
SLX_PSC_P1BB0_TOTAL_ENGINE_TIME                0.000000
DAY                                            0.000000
Length: 451, dtype: float64


## Category Pipeline


In [106]:
# Date transformation
data = extraTrasfromToCategory(data, debug=True)

'''
# Zero Variance removal
data = removeZeroCategory(data)
print ('After all removal pipeline - ',data.shape)
'''

Dtypes are: 
DAY                                  object
MONTH                                object
YEAR                                 object
VFE_0005_VEHICLE_OPERATION_DIGIT1    object
VFE_0006_VEHICLE_OPERATION_DIGIT2    object
VFE_0008_HAS_PTO                     object
dtype: object


"\n# Zero Variance removal\ndata = removeZeroCategory(data)\nprint ('After all removal pipeline - ',data.shape)\n"

In [107]:
'''# First time 
# find all categorical features.
data_cat = data.select_dtypes(include='object')
cat_features_df = embeddingSizeExtract(data_categorical=data_cat)
cat_features_df

# save this embedding dataframe information to a csv file
cat_features_df.to_csv("drive/My Drive/Thesis/data/Embedding_Data/cat_embedding_details.csv", sep=',', index=False)
'''
# load this information from the save csv file
cat_features_df = pd.read_csv("drive/My Drive/Thesis/data/Embedding_Data/cat_embedding_details.csv", sep=',', low_memory=False)
cat_features_df

Feature  Cardinality  Embedding_Size
0                                 DAY            2               1
1                               MONTH           12               6
2                                YEAR            4               2
3             VAR_7MA_FRONT_AXLE_TYPE            2               1
4                 VAR_DPX_ENGINE_TYPE            3               2
..                                ...          ...             ...
69                 VFE_0009_LTVF_TYPE            6               3
70  VFE_0005_VEHICLE_OPERATION_DIGIT1            5               3
71  VFE_0006_VEHICLE_OPERATION_DIGIT2            9               5
72                VFE_0007_CAB_HEIGHT            7               4
73                   VFE_0008_HAS_PTO            2               1

[74 rows x 3 columns]

In [108]:
data.shape

(19954, 451)

(336677, 449)

# Training and Validation Data Extraction
Based on the previous understanding. We are going to have to nature of data
- Numerical Features and Catgorical Features.
The Numerical features have a total of 365 features while the Categorical feature has 72 features.

We will have Training and Validation data.

In [0]:
'''
# extracting the categorical features
# extracting the categorical features
cat_features = cat_features_df['Feature'].values
# numerical features are
num_features = data.select_dtypes(exclude='object').columns

# create a dataframe of the numerical and categorical features and store the result
feature_names_cat = pd.DataFrame(data=None)
feature_names_num = pd.DataFrame(data=None)
feature_names_num['Numerical'] = num_features
feature_names_cat['Categorical'] = cat_features

# save this feature names of categorical and numerical features to a csv files
feature_names_cat.to_csv("drive/My Drive/Thesis/data/Embedding_Data/feature_names_cat.csv", sep=',', index=False)
# save this feature names of categorical and numerical features to a csv files
feature_names_num.to_csv("drive/My Drive/Thesis/data/Embedding_Data/feature_names_num.csv", sep=',', index=False)
'''
# load feature names from db
feature_names_cat = pd.read_csv("drive/My Drive/Thesis/data/Embedding_Data/feature_names_cat.csv", sep=',', low_memory=False)
feature_names_num = pd.read_csv("drive/My Drive/Thesis/data/Embedding_Data/feature_names_num.csv", sep=',', low_memory=False)


In [0]:
# fetch the categorical and numerical features
cat_features = feature_names_cat['Categorical'].values
num_features = feature_names_num['Numerical'].values


In [111]:
num_features.shape

(362,)

In [0]:
from sklearn.model_selection import train_test_split, GroupShuffleSplit
# for cross validation, we can use Group KFold spliting - GroupShuffleSplit


In [0]:
# Function for assisting in spliting the data
def validationSplit(data_in, split=0.2, toShuffle=False):
  # here we will split the data
  # splitting by chassis
  chassis_data = data_in['T_CHASSIS'].unique()
  # splitting the chassis data
  c_train, c_val = train_test_split(chassis_data, test_size=split, random_state=42, shuffle=toShuffle)
  # extracting out the training and testing
  train_data = data_in[data_in['T_CHASSIS'].isin(c_train)].reset_index(drop=True)
  test_data = data_in[data_in['T_CHASSIS'].isin(c_val)].reset_index(drop=True)
  return train_data, test_data


In [0]:
# drop na
data = data.dropna().reset_index(drop=True)

In [0]:
# Function for assisting in spliting the data
def crossValidationSplit(data_in, split=0.2, toShuffle=False):
  # here we will split the data based on their chassis grouping
  chassis_group = data['T_CHASSIS'].values
  ## generating the Group Parameters
  c_split = GroupShuffleSplit(n_splits=2, test_size=0.2, random_state=42)
  '''
  # splitting the data
  for train_idx, test_idx in c_split.split(data.values, groups=chassis_group):
    print("TRAIN:", data.values[train_idx].shape, "TEST:", data.values[test_idx].shape)
  '''
  return c_split

### Train Splitting

In [116]:
data_t, data_v = validationSplit(data, split=0.2)
data_t.shape, data_v.shape

((12884, 451), (2978, 451))

In [0]:
#data_t.to_csv("drive/My Drive/Thesis/data/Embedding_Data/chassis_testing_data.csv", sep=',', index=False)
#data_v.to_csv("drive/My Drive/Thesis/data/Embedding_Data/chassis_training_data.csv", sep=',', index=False)

## Category Summary
In summary, we have a total of 74 categories. They can be used to filter new data sources instead of going through the whole extaction and category pipeline all over again.

## Handling numerical features
We have a total of 362 numerical features

## Embeddings

In order to apply embeddings for out categorical variables. We need to first determine which features are categorical variable and which are not. We need to ensure we caputure all possible cases of the categorical features. 

Okay, then for each categorical variable we need to capture the cardinalty of the feature iteself.

In [0]:
# use tensorflow version 2
%tensorflow_version 2.x

In [0]:
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.layers import Embedding, Input, Reshape, Concatenate, Dense, Flatten, Dropout

from tensorflow.keras.callbacks import TensorBoard
import os
import datetime

from tensorflow.keras.utils import plot_model
from tensorflow.keras.callbacks import EarlyStopping

In [0]:
#from tensorflow.summary import create_file_writer
#  TensorFlow and the TensorBoard HParams plugin:
import tensorflow as tf
from tensorboard.plugins.hparams import api as hp


In [0]:
### Write metadata file for embeddings
def write_metadata(filename, labels):
    """
            Create a metadata file image consisting of sample indices and labels
            :param filename: name of the file to save on disk
            :param shape: tensor of labels
    """
    with open(filename, 'w') as f:
        f.write("Index\tLabel\n")
        for index, label in enumerate(labels):
            f.write("{}\t{}\n".format(index, label))

    print('Metadata file saved in {}'.format(filename))
    

In [0]:
#data_t[cat_features].nunique()

In [123]:
data_t['VFE_0007_CAB_HEIGHT'].unique()

array(['CABH165', 'CABH230', 'CABH215', 'CABH175', 'CABH210', 'Low',
       'CABH225'], dtype=object)

In [0]:
feature_col_labels = pd.factorize(data_t.loc[:, 'VFE_0007_CAB_HEIGHT'].values)[0]
metadata_path = "drive/My Drive/Thesis/data/Embedding_Data/metadata_CAB_HEIGHT.tsv"
#write_metadata(metadata_path, feature_col_labels)

In [0]:
embeddings_meta = {'Embedding_layer_VFE_0007_CAB_HEIGHT':metadata_path}


### Saving and Loading Embeddings

In [0]:
# Goal
'''
Write a function, when given a categorical variable, it will generate the vector representation for that variable.
In order to achieve that, we first need to map very categorical variable to some numerical index value. This value 
has to be unique and not random. 
'''

def transfromForEmbed(data_cat, cat_feature_name='', cat_mapping=''):
  # find the cat_index for all the features
  def findCatIndex(x):
    return int(cat_mapping[cat_mapping[cat_feature_name] == x][cat_feature_name+'_index'].values[0])
  output = data_cat[cat_feature_name].apply(findCatIndex) 
  return output

def embedTransform(model=None, variable='', cat_feature_name='', cat_mapping='', return_type=1):
  # Given a categorical variable, produce the embedded vector
  # Find the index of the categorical variable in the cat_mapping dataframe
  cat_index = int(cat_mapping[cat_mapping[cat_feature_name] == variable][cat_feature_name+'_index'].values[0])
  if return_type == 0:
    return cat_index
  # fetch the embeddings weights
  pre_embedding = "Embedding_layer_"
  embed_layer = model.get_layer(name=pre_embedding+cat_feature_name)
  embedd_vector = embed_layer.get_weights()[0][cat_index]
  return embedd_vector

def createCatMapping(data_cat):
  # this function will create a mapping dataframe
  cat_size = data_cat.shape[1]
  feature_names = data_cat.columns

  cat_mapping = pd.DataFrame()

  col_mapping_list = list()
  col_mapping_cols = list()
  # try for all dataframes
  for i in range(cat_size):
    col_indexs = list(pd.factorize(data_cat.iloc[:, i].unique())[0])
    # save to list
    col_mapping_list.append(list(data_cat.iloc[:, i].unique()))
    col_mapping_list.append(col_indexs)
    # save the column name as well
    col_mapping_cols.append(feature_names[i])
    col_mapping_cols.append(feature_names[i]+'_index')

  # now we need to save the list to a dataframe
  cat_mapping = pd.DataFrame(col_mapping_list)
  # transpose it
  cat_mapping = cat_mapping.transpose()
  # update the columns names
  cat_mapping.columns = col_mapping_cols

  # save to the dataframe
  cat_mapping.to_csv("drive/My Drive/Thesis/data/Embedding_Data/cat_mappings.csv", sep=',', index=False)

#createCatMapping(data[cat_features]) 

In [0]:
# load from file
cat_mappings = pd.read_csv("drive/My Drive/Thesis/data/Embedding_Data/cat_mappings.csv", sep=',', low_memory=False)


In [128]:
data_t['VFE_0007_CAB_HEIGHT'].unique()

array(['CABH165', 'CABH230', 'CABH215', 'CABH175', 'CABH210', 'Low',
       'CABH225'], dtype=object)

In [129]:
# load the embedding model
from tensorflow.keras.models import load_model
# identical to the previous one
embed_model = load_model("drive/My Drive/Thesis/data/Embedding_Data/embedding_model.h5")


In [216]:
# find the embedding vector for feature CABH210 in VFE_0007_CAB_HEIGHT
'''
my_vector1 = embedTransform(embed_model, variable='CABH175', cat_feature_name='VFE_0007_CAB_HEIGHT', cat_mapping=cat_mappings)
my_vector2 = embedTransform(embed_model, variable='CABH165', cat_feature_name='VFE_0007_CAB_HEIGHT', cat_mapping=cat_mappings)
my_vector3 = embedTransform(embed_model, variable='CABH210', cat_feature_name='VFE_0007_CAB_HEIGHT', cat_mapping=cat_mappings)
my_vector4 = embedTransform(embed_model, variable='CABH230', cat_feature_name='VFE_0007_CAB_HEIGHT', cat_mapping=cat_mappings)
my_vector5 = embedTransform(embed_model, variable='Low', cat_feature_name='VFE_0007_CAB_HEIGHT', cat_mapping=cat_mappings)
'''
x1 = output_embeddings[0]
x2 = output_embeddings[1]

# calculating the cosing similarity of two vectors in the orginal space
a_dot = np.dot(x1, x2)
similar = a_dot / (np.linalg.norm(x1) * np.linalg.norm(x2))
similar


0.9994888

In [0]:
# betwen CABH175 and CABH165, i have 0.867487
# betwen CABH175 and CABH210, i have 0.11280868
# betwen CABH175 and CABH230, i have -0.5284594
# betwen CABH175 and Low, i have -0.36939222


In [131]:
ddsad = embed_model.get_layer(name='Embedding_layer_VFE_0007_CAB_HEIGHT')
ddsad.get_weights()

[array([[-0.14905094, -0.11075342, -0.05900112, -0.0247627 ],
        [ 0.02609651, -0.06145609,  0.13505827, -0.00065673],
        [-0.06022814, -0.05502219, -0.04270122, -0.07960243],
        [ 0.21226509, -0.04694704,  0.28343254,  0.00852349],
        [ 0.04596679,  0.11171044,  0.00124936,  0.05315955],
        [ 0.0416726 , -0.0106741 ,  0.12163171, -0.02237656],
        [-0.04100224, -0.0463232 , -0.00980473, -0.0373629 ]],
       dtype=float32)]

### Creating the Keras Embeddings

In [0]:
# Load the TensorBoard notebook extension
%load_ext tensorboard


In [0]:
# Clear any logs from previous runs
!rm -rf ./logs/ 

In [0]:
logdir = os.path.join("logs", datetime.datetime.now().strftime("%Y%m%d-%H%M%S"))
#tensorboard_callback = TensorBoard(logdir, histogram_freq=1, embeddings_freq=1, embeddings_metadata=embeddings_meta, write_images=True)
tensorboard_callback = TensorBoard(logdir, histogram_freq=1, embeddings_freq=1,
                                   write_images=False)


In [0]:
### Creating the Keras Embeddings Mode

def create_model(saved_model=None):
  # IDs representing 1-hot encodings
  # Need to create the input for all features
  cat_embedding_input_layers = []

  cat_reshape_layers = []
  # Embeddings for the first cat feature
  id_feature = 0
  # create an array of embeddings layers
  cat_embedding_layers = []

  # interate through categorical varaibles
  for id_feature in range(cat_features_df.shape[0]):
    # creating the input layer for the embeddings
    #input_layer = Input(shape=(1,), name="Input_layer_"+cat_features_df.loc[id_feature, 'Feature'])
    input_layer = Input(shape=(1,), name="Input_layer_"+cat_features_df.loc[id_feature, 'Feature'])
    cat_embedding_input_layers.append(input_layer)
    # embedding size 
    layer_embedding_size = cat_features_df.loc[id_feature, 'Embedding_Size']
    # create the embedding layers
    embed_weights = saved_model.get_layer(name="Embedding_layer_"+cat_features_df.loc[id_feature, 'Feature']).get_weights()
    #embedded_layer.trainable = False
    embedded_layer = Embedding(input_dim=cat_features_df.loc[id_feature, 'Cardinality'], output_dim=layer_embedding_size, 
                               name="Embedding_layer_"+cat_features_df.loc[id_feature, 'Feature'], input_length = 1, trainable=False, weights=embed_weights)(input_layer)
    cat_embedding_layers.append(embedded_layer)
    # add a reshape of the embedding layers
    reshape_layer = Reshape(target_shape=(layer_embedding_size,))(embedded_layer)
    # appends the rehshape models together
    cat_reshape_layers.append(reshape_layer)

  ### Create A combined embedding layers only
  combined_emb = Concatenate(axis=1, name='combined_embeddings')(cat_reshape_layers)
  
  ### the output of the combine embedding model is 169 vectors

  ### creating a model based on the trained weights
  model_emb = Model(inputs=cat_embedding_input_layers, outputs=combined_emb, name="Categorical_Embeddings")

  # compile the model - for a regression model
  model_emb.compile(optimizer='adam', loss='mean_absolute_error', metrics=['mean_absolute_error'])
  # compile the model - for a classification problem
  # model.compile(loss='binary_crossentropy', optimizer='adam')
  return model_emb

def train_model(model, x_train, y_train, x_test, y_test, epochs, batch_size, early_stop_callback=None, tensorboard_callback=None):
  # fitting the model
  #model.fit(x=x_train, y=y_train, epochs=epochs, batch_size=batch_size, verbose=2, validation_data=(XTest, yTest), callbacks=[early_stop, tensorboard_callback])
  model.fit(x=x_train, y=y_train, epochs=epochs, batch_size=batch_size, verbose=2, callbacks=[early_stop_callback, tensorboard_callback])
  # evaluate the model
  _, metric_mae = model.evaluate(x_test, y_test)
  return metric_mae

In [0]:
## For each run, log an hparams summary with the hyperparameters and final accuracy:
def run(my_model, x_train, y_train, x_test, y_test, epochs, batch_size, early_stop_callback, tensorboard_log):
  model = None
  model = create_model(my_model)
  metric_mae = train_model(model, x_train, y_train, x_test, y_test, epochs, batch_size, early_stop_callback=early_stop_callback, tensorboard_callback=tensorboard_log)
  # do something with the metrics
  return model, metric_mae

In [0]:
model = create_model(saved_model=embed_model)
model.summary()

In [0]:
# plot the model
#plot_model(model=model_emb, to_file='drive/My Drive/Thesis/data/Embedding_Data/model_image.png', show_shapes=True, dpi=300)
#plot_model(model=model_emb, to_file='drive/My Drive/Thesis/data/Embedding_Data/model_new.png', show_shapes=True, dpi=300)

In [0]:
#model_emb.summary()

## Inputs to the Network

We need to create a list of inputs, such that each categorical feature has it's own list of values and the last list will be a 2D array of the continous values. As shown above.

The size of the list of inputs : ip = n_cat + 1

Remember for each of the embedding network we had set input-size =1 we are taking 1 value each from all the list (except the last list) and sending it to the combined network for training. For the last list, each value itself is a list having the other columns values, and this is sent to the models_rest network

In [0]:
from sklearn.preprocessing import RobustScaler

In [0]:
def transfromForEmbed(data_cat, cat_feature_name='', cat_mapping='', type=None):
  # find the cat_index for all the features
  def findCatIndex(x):
    return int(cat_mapping[cat_mapping[cat_feature_name] == x][cat_feature_name+'_index'].values[0])
  if type is not None:
    output = findCatIndex(data_cat[cat_feature_name])
    return np.array(output)
  else:
    output = data_cat[cat_feature_name].apply(findCatIndex) 
    return output.values

# function to reformat the input data into the necessary input data for our network to be trained on.
def network_input_process(data_num=None, data_cols=None, y_output=None, input_type=1):
  # create a list of inputs. 
  try:
    cat_size = data_cols.shape[1]
    columns_name = data_cols.columns
    col_type = None
  except:
    cat_size = data_cols.shape[0]
    columns_name = data_cols.index
    col_type = 1

  # define the size of the network list inputs
  if input_type==1:
    network_inputs = [None] * (cat_size + 1)
  else:
    network_inputs = [None] * (cat_size)

  # add inputs to the list
  # for the categorical inputs

  for i in range(cat_size):
    #network_inputs[i] = pd.factorize(data_cols.iloc[:, i].values)[0]
    # map each categorical variable to their respective index values
    network_inputs[i] = transfromForEmbed(data_cols, cat_feature_name=columns_name[i], cat_mapping=cat_mappings, type=col_type)

  if input_type == 1:
    # for the continuous value inputs
    network_inputs[-1] = data_num.iloc[:, :].values
    # that should be all
    return network_inputs, y_output.values
  else:
    return network_inputs

In [0]:
# creating the training inputs
scaler = RobustScaler()
# scale and fit
scaler.fit(data_t.loc[:, num_features].values)
data_t.loc[:, num_features] = scaler.transform(data_t.loc[:, num_features].values)
data_v.loc[:, num_features] = scaler.transform(data_v.loc[:, num_features].values)

Xtrain, yTrain = network_input_process(data_num=data_t[num_features], data_cols=data_t[cat_features], y_output=data_t['RUL'])
XTest, yTest = network_input_process(data_num=data_v[num_features], data_cols=data_v[cat_features], y_output=data_v['RUL'])

## Training the Network

In [0]:
early_stop = EarlyStopping(monitor='mean_absolute_error', mode='min', verbose=1, patience=3)

In [0]:
# fitting the model
# model_emb.fit(x=Xtrain, y=yTrain, epochs=600, batch_size=64, verbose=2, validation_data=(XTest, yTest), callbacks=[early_stop, tensorboard_callback])


In [0]:
# remove old logs
!rm -rf ./logs/ 

In [148]:
#train the model
my_model, metric_result = run(embed_model, Xtrain, yTrain, XTest, yTest, 100, 64, early_stop, tensorboard_log=tensorboard_callback)


Epoch 1/100
202/202 - 1s - loss: 395.2553 - mean_absolute_error: 395.0915
Epoch 2/100
202/202 - 1s - loss: 394.8163 - mean_absolute_error: 395.0915
Epoch 3/100
202/202 - 1s - loss: 394.9405 - mean_absolute_error: 395.0914
Epoch 4/100
202/202 - 1s - loss: 394.9830 - mean_absolute_error: 395.0913
Epoch 5/100
202/202 - 1s - loss: 395.0585 - mean_absolute_error: 395.0915
Epoch 6/100
202/202 - 1s - loss: 395.0342 - mean_absolute_error: 395.0913
Epoch 7/100
202/202 - 1s - loss: 394.6909 - mean_absolute_error: 395.0914
Epoch 00007: early stopping
94/94 [==============================] - 0s 2ms/step - loss: 360.4678 - mean_absolute_error: 363.9782


In [0]:
data_sample

## Testing Embeddings


In [0]:
data_sample = data_v[cat_features].iloc[100:150,:]
sample_data_point = network_input_process(data_cols=data_sample, input_type=0)

In [0]:
sample_data_point

In [207]:
len(sample_data_point)

74

In [0]:
output_embeddings = my_model.predict(sample_data_point)

In [214]:
output_embeddings.shape

(2, 169)

In [0]:
print ('result - ', metric_result)
# see model summary
#my_model.summary()

# save the model
#my_model.save("drive/My Drive/Thesis/data/Embedding_Data/embedding_model.h5")

In [0]:
my_layerss = my_model.get_layer(index=111)

ddsad = my_model.get_layer(name='Embedding_layer_VFE_0007_CAB_HEIGHT')

In [0]:
data_t['VFE_0007_CAB_HEIGHT'].unique()

In [0]:
ddsad.get_config()

In [0]:
ddsad.embeddings

In [0]:
ddsad.get_weights()[0][6]

'''
First iteration 
using 512 - 256 - 128
###
error 202/202 - 3s - loss: 4807.8838 - val_loss: 263093.3125
Epoch 71/600
202/202 - 3s - loss: 4637.8784 - val_loss: 352122.9062
Epoch 00071: early stopping
mean-absolute error - 245.14920879082842
e_var = -4.4
/---------------------
using 1028 - 512 - 128


----------------------
202/202 - 4s - loss: 828.3008 - val_loss: 443548.0000
-----------------------
Epoch 57/600
202/202 - 5s - loss: 5574.1982 - val_loss: 124869.9609
------------------------
With ELU : 202/202 - 4s - loss: 2499.8364 - val_loss: 66855.7031



In [0]:
# plot the loss
lossess = pd.DataFrame(my_model.history.history)

In [0]:
lossess

In [0]:
lossess.plot(figsize=(12,6))

In [0]:
lossess['val_loss'].plot(figsize=(12,6))

In [0]:
# evaluating our model
np.sqrt(model_emb.evaluate(x=Xtrain, y=yTrain, verbose=1))

In [0]:
# evaluating our model
np.sqrt(model_emb.evaluate(x=XTest, y=yTest, verbose=1))

In [0]:
# Predicting the model
test_predicitons = model_emb.predict(XTest)
test_predicitons.shape

In [0]:
# plotting our prediction result
test_predictions = pd.Series(test_predicitons.reshape(2978,))
pred_df = pd.DataFrame(yTest, columns=['True RUL'])
pred_df = pd.concat([pred_df, test_predictions], axis=1)
pred_df.columns = ['True RUL', 'Predicted RUL']

In [0]:
pred_df.head(10)

In [0]:
from sklearn.metrics import mean_absolute_error, explained_variance_score

In [0]:
# mean absolute error for our predictions
mean_absolute_error(pred_df['True RUL'], pred_df['Predicted RUL'])

In [0]:
# so is our error good or bad?
data['RUL'].describe()

In [0]:
# explained variance score
explained_variance_score(pred_df['True RUL'].values, pred_df['Predicted RUL'].values)

In [0]:
!kill 1142

In [0]:
###
# running tensorboard
tensorboard --logdir logs